### PR TITLE
refactor: PHP 8.0 constructor promotion

### PR DESCRIPTION
**Pull request by kenjis**
Needs #6923

**Description**
See #6921

Use constructor promotion.
This may contain breaking changes.

**Checklist:**
- [x] Securely signed commits
- [ ] Component(s) with PHPDoc blocks, only if necessary or adds value
- [ ] Unit testing, with >80% coverage
- [ ] User guide updated
- [x] Conforms to style guide

### CODE REVIEW

**Comment by kenjis** [comment]
This is a breaking chage:
https://3v4l.org/BrHEP

**Comment by samsonasik** [comment]
Yes, that seems need to be skipped

File: system/Autoloader/FileLocator.php [hunk 1]
```diff
@@ -18,15 +18,11 @@
...
+     * @param Autoloader $autoloader The Autoloader to use.
      */
-    protected $autoloader;
-
-    public function __construct(Autoloader $autoloader)
-    {
-        $this->autoloader = $autoloader;
+    public function __construct(
+        protected Autoloader $autoloader
+    ) {
```
Comment (kenjis) [inline system/Autoloader/FileLocator.php:25]: For example, if a dev extends `FileLocator` and it has `protected $autoloader;`,
this change is a breaking change.
https://3v4l.org/NhqDb

Comment (samsonasik) [reply system/Autoloader/FileLocator.php:25]: There are possible options:

- make a note about property re-definition possible breaking change without add type
- wait to push to 4.4 only

Comment (samsonasik) [reply system/Autoloader/FileLocator.php:25]: Another way;: make Rector extension to help user migrating to new definition.

Comment (kenjis) [reply system/Autoloader/FileLocator.php:25]: Removing type in the constructor makes no errors (I don't know why), 
https://3v4l.org/2A85L
but I don't want to remove already declared types in constructors.

Comment (samsonasik) [reply system/Autoloader/FileLocator.php:25]: Better way: make a patch to rector to:

- skip non-typed property for protected/public property, except define `INLINE_PUBLIC` config that allow bc break on non-final class, same like TypedPropertyRector way

That's just quick guessing, probably there is a better way.

**Comment by samsonasik** [comment]
@kenjis some rules like `StrEndsWithRector`, `StrStartsWithRector`, `StrContainsRector` seems easy merge 👍 , define small PRs and iterate to next rules when possible.

File: system/Autoloader/FileLocator.php (hunk 1 above)
Comment (samsonasik) [reply system/Autoloader/FileLocator.php:25]: @kenjis here possibly the solution, I create PR to make it configurable `INLINE_PUBLIC` default false:

- https://github.com/rectorphp/rector-src/pull/3126

**Comment by kenjis** [comment]
If we drop PHP 7.4 in v4.4, I would like to type missing public/protected properties with constructor promotion.
These will be breaking changes, though.

**Comment by samsonasik** [comment]
Yes, I agree 👍

File: system/Autoloader/FileLocator.php (hunk 1 above)
Comment (MGatner) [reply system/Autoloader/FileLocator.php:25]: This seems the best approach to me. Glad your patch was already accepted!

**Pull request closed without merging**